### PR TITLE
avoid a couple of warnings in `polyfit`

### PR DESCRIPTION
**Pull request by keewis**
- [x] towards #8844
---
- replace `numpy.core.finfo` with `numpy.finfo`
- add `dtype` and `copy` parameters to all definitions of `__array__`

### CODE REVIEW

File: xarray/namedarray/_typing.py [hunk 1]
```diff
@@ -153,13 +153,15 @@ def __getitem__(
...
+        self, dtype: None = ..., copy: None = ..., /
+    ) -> np.ndarray[Any, _DType_co]: ...
 
     @overload
-    def __array__(self, dtype: _DType, /) -> np.ndarray[Any, _DType]: ...
+    def __array__(self, dtype: _DType, copy: bool, /) -> np.ndarray[Any, _DType]: ...
 
     def __array__(
-        self, dtype: _DType | None = ..., /
+        self, dtype: _DType | None = ..., copy: bool | None = None, /
```
Comment (keewis) [inline xarray/namedarray/_typing.py:164]: @Illviljan (or anyone else), I might need some help to get the typing right... basically, `__array__` is supposed to have a new kwarg, `copy`, which is either a `bool` or `None`.

**Comment by keewis** [comment]
I've removed all the forwarding of `copy` to `asarray` since that doesn't have the kwarg until `numpy>=2`.

If we can get the type checking to pass this should be ready for reviews and merging.

File: xarray/namedarray/_typing.py (hunk 1 above)
Comment (Illviljan) [reply xarray/namedarray/_typing.py]: There's more places with `__array__` to fix:
```
xarray/core/indexing.py:472: note:     Got:
xarray/core/indexing.py:472: note:         @overload
xarray/core/indexing.py:472: note:         def __array__(self, None = ..., /) -> ndarray[Any, dtype[generic]]
xarray/core/indexing.py:472: note:         @overload
xarray/core/indexing.py:472: note:         def [_DType <: dtype[Any]] __array__(self, _DType, /) -> ndarray[Any, _DType]
```

Comment (Illviljan) [reply xarray/namedarray/_typing.py]: numpy's type hints:
https://github.com/numpy/numpy/blob/ab7649fe2ed8f0f0260322d66631b8dfab57deff/numpy/__init__.pyi#L1466-L1473

**Comment by keewis** [comment]
okay, so I believe I've fixed most of the issues. However, since the `copy` kwarg is new in `numpy>=2`, `mypy` will fail to type check properly for versions of `numpy<2`. I have no idea how to keep backwards compatibility in type checking, so I'll give up on that. If anyone has an idea, feel free to push directly to my branch.

**Comment by max-sixty** [comment]
> However, since the `copy` kwarg is new in `numpy>=2`, `mypy` will fail to type check properly for versions of `numpy<2`. I have no idea how to keep backwards compatibility in type checking, so I'll give up on that. 

Definitely the correct call imo!

**Comment by keewis** [comment]
> But can we ensure the `mypy` env has the latest version so that it passes in CI?

The issue here is that this is the upcoming `numpy` 2.0 release, of which at the moment there is only a release candidate. So while upgrading the 2 `mypy` CIs might work, someone will have to check the upstream-dev `mypy` CI to see if there's anything to fix (there definitely is, though it would probably be best to wait with that until we figured out what to do with the runtime errors).

In general, though, I think we need to figure out what to do with backwards compatibility in typing.

**Comment by max-sixty** [comment]
Would it be reasonable to lock the mypy tests at numpy <2, and then when numpy>=2 comes out, we do a single PR which upgrades it to numpy>=2? And then we can merge this then?

I think trying to fix mypy for multiple dependency versions might be too hard (I already find it difficult for a single version!)

**Comment by keewis** [comment]
probably. However, I don't know enough about typing to get that to work. So, if anyone has ideas, feel free to push to this PR.

**Comment by max-sixty** [comment]
(OK — just tbc — my suggestion is not to change the typing. But instead to lock the CI dependencies for mypy at numpy <2 until numpy 2 is released, at which point we lock to numpy > 2. And then to merge anything that requires numpy 2 typing, which may include this PR...)

**Comment by keewis** [comment]
you mean, let's try reverting the typing and see if everything passes? I remember changing the typing because the CI [failed](https://github.com/pydata/xarray/actions/runs/8673262512/job/23784430531), so as far as I can tell (with my limited knowledge of typing) at least _some_ changes are necessary.

**Comment by max-sixty** [comment]
I'm saying that if the typing is proving difficult, we should abandon the goal of mypy passing with both numpy 1 and numpy 2 as dependencies.

For tests, we should really try to pass with both numpy 1 & 2, because we can easily have if/else branches, and it affects people running the code. 

But with typing it's both more difficult and less useful...

**Comment by keewis** [comment]
makes sense. What I'm saying, though, is that without any changes at all to the typing, `mypy` won't pass either. So we might have to find a way to be runtime-compatible with `numpy>=2` without attempting to be typing-compatible.

**Comment by max-sixty** [comment]
> find a way to be runtime-compatible with `numpy>=2` without attempting to be typing-compatible (but how to do that I don't know).

Either our trusty `type: ignore` or to wait for numpy 2...?

**Comment by keewis** [comment]
Since I really have no idea how to fix the typing issues (ignore comments won't work because we usually don't call the protocol ourselves), I've decided to ignore the warning in all our CI for now, and punt the change to the signature in most cases to a future PR. Hopefully this allows us to move forward with this PR.

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -8922,9 +8922,7 @@ def polyfit(
         lhs = np.vander(x, order)
 
         if rcond is None:
-            rcond = (
-                x.shape[0] * np.core.finfo(x.dtype).eps  # type: ignore[attr-defined]
-            )
+            rcond = x.shape[0] * np.finfo(x.dtype).eps  # type: ignore[attr-defined]
```
Comment (dcherian) [inline xarray/core/dataset.py:8925]:
```suggestion
            rcond = x.shape[0] * np.finfo(x.dtype).eps
```

**Comment by keewis** [comment]
looks like this works now? Should we merge?

**Comment by dcherian** [comment]
Thanks for working through this @keewis Can you open an issue please?

**Pull request merged**